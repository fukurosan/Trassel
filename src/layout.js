import { initializeNodesAndEdges } from "./util/initializer"
import Loop from "./loop"
import Quadtree from "./util/quadtree"

/**
 * Main layout class
 */
export default class Layout {
	/**
	 * @param {import("./model/ibasicnode").IBasicNode[]=} nodes - Initial nodes
	 * @param {import("./model/ibasicedge").IBasicEdge[]=} edges - Initial edges
	 * @param {import("./model/ioptions").ILayoutOptions} options - options
	 */
	constructor(nodes = [], edges = [], options = {}) {
		this.nodes = nodes
		this.edges = edges
		this.alpha = options.alpha || 1
		this.alphaMin = options.alphaMin || 0.001
		this.alphaDecay = options.alphaDecay || 1 - Math.pow(this.alphaMin, 1 / 300)
		this.alphaTarget = options.alphaTarget || 0
		this.velocityDecay = options.velocityDecay || 0.6
		/** @type {Map<string, import("./model/ilayoutcomponentobject").ILayoutComponentObject>} */
		this.components = new Map()
		this.listeners = new Map([
			["layoutloopstart", new Set()],
			["layoutupdate", new Set()],
			["layoutloopend", new Set()]
		])
		this.loop = new Loop(this.runLoop.bind(this), options.updateCap ? options.updateCap : 60)
		this.initializeNodesAndEdges()
		this.quadtree = new Quadtree(this.nodes)
		this.isAnimating = false
	}

	/**
	 * Registers an event listener
	 * @param {string} name - Event name to listen for
	 * @param {() => any} fn - Callback on event
	 */
	on(name, fn) {
		if (!this.listeners.has(name)) {
			console.error(`No such event name: ${name}`)
		}
		this.listeners.get(name).add(fn)
	}

	triggerEvent(name) {
		this.listeners.get(name).forEach(fn => fn())
	}

	/**
	 * This is the main loop function.
	 * Each time the loop instance triggers an update this will execute.
	 */
	runLoop() {
		this.update()
		if (this.alpha < this.alphaMin) {
			this.loop.stop()
			this.triggerEvent("layoutloopend")
		}
	}

	updateNodesAndEdges(nodes, edges) {
		this.nodes = nodes
		this.edges = edges
		this.initializeNodesAndEdges()
		this.components.forEach(component => this.initializeComponent(component))
	}

	initializeNodesAndEdges() {
		initializeNodesAndEdges(this.nodes, this.edges)
		this.quadtree = new Quadtree(this.nodes)
	}

	initializeComponent(component) {
		let nodes = this.nodes
		let edges = this.edges
		if (component.nodeBindings) {
			nodes = this.nodes.filter(node => component.nodeBindings(node))
		}
		if (component.edgeBindings) {
			edges = this.edges.filter(edge => component.edgeBindings(edge))
		}
		component.instance.initialize(nodes, edges, { quadtree: this.quadtree, remove: () => this.removeComponent(component.id) })
		return component
	}

	/**
	 * Starts the layout loop
	 */
	start() {
		this.triggerEvent("layoutloopstart")
		this.loop.start()
	}

	/**
	 * Stops the layout loop
	 */
	stop() {
		this.loop.stop()
		this.triggerEvent("layoutloopend")
	}

	/**
	 * Sets the update cap (per second) for the layout loop
	 * @param {number} newCap - new cap
	 */
	setUpdateCap(newCap) {
		this.loop.setUpdateCap(newCap)
	}

	/**
	 * Adds a component to the layout
	 * @param {string} id
	 * @param {import("./model/ilayoutcomponent").ILayoutComponent} component - A layout component compatible class instance
	 * @param {(any) => boolean=} nodeBindings - Function that computes if a node should be affected by the component. Blank means true for all.
	 * @param {(any) => boolean=} edgeBindings - Function that computes if an edge should be affected by the component. Blank means true for all.
	 * @returns {Layout} - this
	 */
	addLayoutComponent(id, component, nodeBindings = null, edgeBindings = null) {
		if (this.components.has(id)) {
			throw new Error("Component already exists: " + id)
		}
		const componentObject = {
			id,
			instance: component,
			nodeBindings,
			edgeBindings
		}
		this.initializeComponent(componentObject)
		this.components.set(id, componentObject)
		return this
	}

	/**
	 * Removes a compnent with the specified ID
	 * @param {string} id
	 */
	removeComponent(id) {
		if (this.components.has(id)) {
			this.components.get(id).instance.dismount()
			this.components.delete(id)
		}
	}

	/**
	 * Finds the node closest to the provided coordinates
	 * @param {number} x
	 * @param {number} y
	 * @returns {any} - The node
	 */
	findClosestNodeByCoordinates(x, y) {
		let closest
		let radius = Infinity
		for (let i = 0; i < this.nodes.length; ++i) {
			const node = this.nodes[i]
			const distanceX = x - node.x
			const distanceY = y - node.y
			const distanceSquared = distanceX * distanceX + distanceY * distanceY
			if (distanceSquared < radius) {
				closest = node
				radius = distanceSquared
			}
		}
		return closest
	}

	/**
	 * Animates nodes from source positions to target positions within a duration provided.
	 * This function can be used to transition the graph between states or layouts.
	 * Once triggered the animation cannot be stopped. All other updates and components will be frozen until the animation completes.
	 * There should *never* be more than one animation running simultaneously.
	 * @param {import("./model/itargetnodestate").ITargetNodeState[]} targetNodeStates
	 * @param {number} duration - Animation duration in milliseconds
	 * @param {boolean} shouldFixateOnEnd - If true then the graph will fixate the nodes when the animation ends
	 */
	animateState(targetNodeStates = [], duration = 300, shouldFixateOnEnd = false) {
		if (!targetNodeStates.length) return
		this.nodeMap = this.nodes.reduce((acc, node) => {
			acc[node.id] = node
			return acc
		}, {})
		targetNodeStates.forEach(state => {
			state.sourceX = isNaN(state.sourceX) ? this.nodeMap[state.id].x : state.sourceX
			state.sourceY = isNaN(state.sourceY) ? this.nodeMap[state.id].y : state.sourceY
		})
		const startTime = Date.now()
		const loop = new Loop(() => {
			const deltaTime = Date.now() - startTime
			const percentOfAnimation = Math.min(deltaTime / duration, 100)
			targetNodeStates.forEach(nodeState => {
				const node = this.nodeMap[nodeState.id]
				node.x = nodeState.sourceX + (nodeState.targetX - nodeState.sourceX) * percentOfAnimation
				node.y = nodeState.sourceY + (nodeState.targetY - nodeState.sourceY) * percentOfAnimation
			})
			if (deltaTime > duration) {
				loop.stop()
				this.isAnimating = false
				if (shouldFixateOnEnd) {
					targetNodeStates.forEach(nodeState => {
						const node = this.nodeMap[nodeState.id]
						node.fx = node.x
						node.fy = node.y
					})
				}
			}
			this.triggerEvent("layoutupdate")
		}, Infinity)
		this.isAnimating = true
		loop.start()
	}

	/**
	 * Main update function.
	 * This executes all components in the layout and computes node positions.
	 * Note that the update function can be executed without the looper.
	 * @param {boolean} sendEvent - Should an update event be fired?
	 */
	update(sendEvent = true) {
		if (this.isAnimating) {
			return
		}
		this.alpha += (this.alphaTarget - this.alpha) * this.alphaDecay
		for (const [, component] of this.components.entries()) {
			component.instance.execute(this.alpha)
		}
		for (let i = 0; i < this.nodes.length; i++) {
			const node = this.nodes[i]
			if (node.fx == null) {
				node.vx *= this.velocityDecay
				node.x += node.vx
			} else {
				node.x = node.fx
				node.vx = 0
			}
			if (node.fy == null) {
				node.vy *= this.velocityDecay
				node.y += node.vy
			} else {
				node.y = node.fy
				node.vy = 0
			}
		}
		this.iteration = this.iteration ? this.iteration + 1 : 1
		this.quadtree.update()
		sendEvent && this.triggerEvent("layoutupdate")
	}
}
