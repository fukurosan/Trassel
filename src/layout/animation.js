import LayoutComponent from "./layoutcomponent"

/**
 * Moves a node from its current position to a provided destination
 * Note that this component uses fx and fy, which could create conflicts with things like pinning mechanisms
 * targetX and targetY coordinates can be set on specific nodes, or component global values can be provided.
 * @param {number=} xDestination - Destination X coordinate
 * @param {number=} yDestination - Destination Y coordinate
 * @param {number=} strength - The strength of the pull
 * @param {number=} removeForceOnDestination - If true the component will be removed from the layout when animation is completed
 */
export default class Animation extends LayoutComponent {
	constructor(xDestination = 0, yDestination = 0, strength = 1, removeForceOnDestination = true) {
		super()
		this.xDestination = xDestination
		this.yDestination = yDestination
		this.strength = strength
		this.removeForceOnDestination = removeForceOnDestination
	}

	initialize(...args) {
		super.initialize(...args)
	}

	execute(alpha) {
		let node
		let xDestination
		let yDestination
		let arrived = 0
		for (let i = 0; i < this.nodes.length; i++) {
			node = this.nodes[i]
			xDestination = node.targetX ? node.targetX : this.xDestination
			yDestination = node.targetY ? node.targetY : this.yDestination
			!node.fx && (node.fx = node.x)
			!node.fy && (node.fy = node.y)
			const xMovement = ((xDestination - node.x) * this.strength * alpha) / 2
			const yMovement = ((yDestination - node.y) * this.strength * alpha) / 2
			xMovement > 1 ? (node.fx += xMovement) : (node.fx = xDestination)
			yMovement > 1 ? (node.fy += yMovement) : (node.fy = yDestination)
			node.fx === xDestination && node.fy === yDestination && arrived++
		}
		if (this.removeForceOnDestination && arrived === this.nodes.length) {
			for (let i = 0; i < this.nodes.length; i++) {
				node = this.nodes[i]
				node.x = node.fx
				node.y = node.fy
				node.vx = 0
				node.vy = 0
				node.fx = undefined
				node.fy = undefined
			}
			this.utils.remove()
		}
	}
}
