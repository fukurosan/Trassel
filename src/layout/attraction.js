import LayoutComponent from "./layoutcomponent"

/**
 * Creates a gravitational force that pulls the x or y axis of nodes towards a given coordinate
 * Two sets of attraction components can commonly be used to create a gravitational force towards a center area
 * This can also be use to draw nodes away from a given point, if a negative strength is provided.
 * @param {number=} isHorizontal - true = x, false = y
 * @param {number=} coordinate - the center coordinate of the component
 * @param {number=} strength - The strength of the pull
 */
export default class Attraction extends LayoutComponent {
	constructor(isHorizontal = true, coordinate = 0, strength = 0.05) {
		super()
		this.isHorizontal = isHorizontal
		this.coordinate = coordinate
		this.strength = strength
	}

	initialize(...args) {
		super.initialize(...args)
	}

	execute(alpha) {
		let node
		for (let i = 0; i < this.nodes.length; i++) {
			node = this.nodes[i]
			node[this.isHorizontal ? "vx" : "vy"] += (this.coordinate - node[this.isHorizontal ? "x" : "y"]) * this.strength * alpha
		}
	}
}
