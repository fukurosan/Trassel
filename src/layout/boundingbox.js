import LayoutComponent from "./layoutcomponent"

/**
 * Creates a rectangular bounding box that stops nodes from leaving it
 * @param {number=} width - Width of the box. If not set will be determined by the sizes and amounts of the nodes
 * @param {number=} height - Height of the box. If not set will be determined by the sizes and amounts of the nodes
 */
export default class BoundingBox extends LayoutComponent {
	constructor(width = null, height = null) {
		super()
		this.width = width
		this.height = height
		this.computedWidth = width
		this.computedHeight = height
		this.multiplier = 5
	}

	getWidth(node) {
		return node.width ? node.width : node.radius * 2
	}

	getHeight(node) {
		return node.height ? node.height : node.radius * 2
	}

	initialize(...args) {
		super.initialize(...args)
		if (!this.width || !this.height) {
			let size = 0
			let node
			for (let i = 0; i < this.nodes.length; i++) {
				node = this.nodes[i]
				size += Math.max(size, this.getWidth(node) * this.multiplier, this.getHeight(node) * this.multiplier)
			}
			if (!this.width && !this.height) {
				this.computedWidth = size / 2
				this.computedHeight = size / 2
			} else if (!this.width) {
				this.computedWidth = size - this.height > 0 ? size - this.height : 100
			} else if (!this.height) {
				this.computedHeight = size - this.width > 0 ? size - this.width : 100
			}
		} else {
			this.computedWidth = this.width
			this.computedHeight = this.height
		}
		this.computedWidth = this.computedWidth / 2
		this.computedHeight = this.computedHeight / 2
	}

	execute() {
		let node
		for (let i = 0; i < this.nodes.length; i++) {
			node = this.nodes[i]
			if (node.x < -this.computedWidth) {
				node.x = -this.computedWidth
			} else if (node.x > this.computedWidth) {
				node.x = this.computedWidth
			}
			if (node.y < -this.computedHeight) {
				node.y = -this.computedHeight
			} else if (node.y > this.computedHeight) {
				node.y = this.computedHeight
			}
		}
	}
}
