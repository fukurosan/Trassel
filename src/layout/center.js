import LayoutComponent from "./layoutcomponent"

/**
 * Uses the average of all node positions to create a center force that stops nodes from floating away
 * @param {number=} x - x coordinate for the component
 * @param {number=} y - y coordinate for the component
 * @param {number=} strength - Strength of the force
 */
export default class Center extends LayoutComponent {
	constructor(x = 0, y = 0, strength = 1) {
		super()
		this.x = x
		this.y = y
		this.strength = strength
	}

	initialize(...args) {
		super.initialize(...args)
	}

	execute() {
		let allX = 0
		let allY = 0
		for (let i = 0; i < this.nodes.length; i++) {
			const node = this.nodes[i]
			allX += node.x
			allY += node.y
		}
		const xAdjustment = (allX / this.nodes.length - this.x) * this.strength
		const yAdjustment = (allY / this.nodes.length - this.y) * this.strength
		for (let i = 0; i < this.nodes.length; i++) {
			const node = this.nodes[i]
			node.x -= xAdjustment
			node.y -= yAdjustment
		}
	}
}
