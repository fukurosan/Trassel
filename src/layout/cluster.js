import LayoutComponent from "./layoutcomponent"

/**
 * Creates a cluster component that draws a set of nodes together
 * @param {number=} strength - How strong should the pull be? (0-1)
 */
export default class Cluster extends LayoutComponent {
	constructor(strength = 0.7) {
		super()
		this.strength = strength
	}

	initialize(...args) {
		super.initialize(...args)
	}

	execute(alpha) {
		let x = 0
		let y = 0
		let z = 0
		let node
		for (let i = 0; i < this.nodes.length; i++) {
			node = this.nodes[i]
			const mass = node.mass ** 2
			x += node.x * mass
			y += node.y * mass
			z += mass
		}
		const centerX = x / z
		const centerY = y / z
		const force = alpha * this.strength
		for (let i = 0; i < this.nodes.length; i++) {
			node = this.nodes[i]
			node.vx -= (node.x - centerX) * force
			node.vy -= (node.y - centerY) * force
		}
	}
}
