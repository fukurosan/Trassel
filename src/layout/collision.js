import LayoutComponent from "./layoutcomponent"

/**
 * Collision component that stops nodes from colliding (and thus overlapping)
 * Note that this component considers all nodes to be circles. If a node is square then the maximum measurement will be considered the diameter.
 * @param {number=} strength - The strength of the collision repulsion
 * @param {number=} radiusPadding - Padding that will be added to all radiuses
 */
export default class Collision extends LayoutComponent {
	constructor(strength = 1, radiusPadding = 5) {
		super()
		this.strength = strength
		this.radiusPadding = radiusPadding
	}

	initialize(...args) {
		super.initialize(...args)
	}

	execute() {
		//Ensure radiuses have been recorded
		this.utils.quadtree.computeLargestRadius(this.radiusPadding)

		let node
		let computedNodeRadius
		for (let i = 0; i < this.nodes.length; i++) {
			node = this.nodes[i]
			computedNodeRadius = node.radius + this.radiusPadding
			this.utils.quadtree.traverseTopBottom((quadNode, xStart, yStart, xEnd, yEnd) => {
				const data = quadNode.entity
				const otherNodeRadius = quadNode.radius + this.radiusPadding
				const combinedRadius = computedNodeRadius + otherNodeRadius
				if (data) {
					//This ensures we do not evaluate the same node combination twice
					if (data.index >= node.index) {
						return
					}
					//Compute the distance between the nodes
					let x = node.x - data.x
					let y = node.y - data.y
					let distanceSquared = x * x + y * y
					//if x^2 + y^2 < r^2 then the two circles are colliding
					if (distanceSquared < combinedRadius * combinedRadius) {
						//we ensure x and y cannot be 0 to guard against 0 divisional error
						if (x === 0) {
							x = this.randomize()
							distanceSquared += x * x
						}
						if (y === 0) {
							y = this.randomize()
							distanceSquared += y * y
						}
						//Compute the actual distance and the force by which the nodes should be adjusted
						const distance = Math.sqrt(distanceSquared)
						const force = ((combinedRadius - distance) / distance) * this.strength
						const xForce = x * force
						const yForce = y * force

						//Compute how much of the force should be applied to node 1 and to node 2 respectively
						const otherNodeRadiusSquared = otherNodeRadius * otherNodeRadius
						const currentNodeRadiusSquared = computedNodeRadius * computedNodeRadius
						const otherRadiusRatio = otherNodeRadiusSquared / (currentNodeRadiusSquared + otherNodeRadiusSquared)
						//Apply force
						node.vx += xForce * otherRadiusRatio
						node.vy += yForce * otherRadiusRatio
						const currentNodeRatio = 1 - otherRadiusRatio
						data.vx -= xForce * currentNodeRatio
						data.vy -= yForce * currentNodeRatio
					}
					return
				}
				return xStart > node.x + combinedRadius || xEnd < node.x - combinedRadius || yStart > node.y + combinedRadius || yEnd < node.y - combinedRadius
			})
		}
	}
}
