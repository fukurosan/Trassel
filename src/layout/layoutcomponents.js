import Animation from "./animation"
import Attraction from "./attraction"
import BoundingBox from "./boundingbox"
import Center from "./center"
import Cluster from "./cluster"
import Collision from "./collision"
import D3Adapter from "./d3adapter"
import Fan from "./fan"
import LayoutComponent from "./layoutcomponent"
import Grid from "./grid"
import Hierarchy from "./hierarchy"
import Link from "./link"
import Matrix from "./matrix"
import NBody from "./nbody"
import Radial from "./radial"

export { Animation, Attraction, BoundingBox, Center, Cluster, Collision, D3Adapter, Fan, LayoutComponent, Grid, Hierarchy, Link, Matrix, NBody, Radial }
