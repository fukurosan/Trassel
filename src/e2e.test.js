import { Attraction, Collision, Link, NBody } from "./layout/layoutcomponents"
import Layout from "./layout"
import Graph from "./graph"

describe("End-to-End Tests", () => {
	const baseNodes = JSON.stringify([
		{ id: "n1" },
		{ id: "n2" },
		{ id: "n3" },
		{ id: "n4" },
		{ id: "n5" },
		{ id: "n6" },
		{ id: "n7" },
		{ id: "n8" },
		{ id: "n9" },
		{ id: "n10" }
	])
	const baseEdges = JSON.stringify([
		{ sourceNode: "n1", targetNode: "n2" },
		{ sourceNode: "n1", targetNode: "n3" },
		{ sourceNode: "n1", targetNode: "n4" },
		{ sourceNode: "n2", targetNode: "n5" },
		{ sourceNode: "n2", targetNode: "n5" },
		{ sourceNode: "n5", targetNode: "n6" },
		{ sourceNode: "n5", targetNode: "n6" },
		{ sourceNode: "n6", targetNode: "n1" },
		{ sourceNode: "n8", targetNode: "n9" },
		{ sourceNode: "n9", targetNode: "n10" }
	])
	let nodes
	let edges
	beforeEach(() => {
		nodes = JSON.parse(baseNodes)
		edges = JSON.parse(baseEdges)
	})

	it("Typical layout loop works", async () => {
		const layout = new Layout(nodes, edges, { updateCap: Infinity })
		layout.addLayoutComponent("collide", new Collision())
		layout.addLayoutComponent("manybody", new NBody())
		layout.addLayoutComponent("x", new Attraction(true))
		layout.addLayoutComponent("y", new Attraction(false))
		layout.addLayoutComponent("link", new Link())
		let updateCount = 0
		layout.on("layoutupdate", () => {
			updateCount++
		})
		let resolve
		const promise = new Promise(promiseResolution => (resolve = promiseResolution))
		layout.on("layoutloopend", () => {
			resolve()
		})
		layout.start()
		await promise
		expect(updateCount).toEqual(300)
		expect(nodes).toMatchSnapshot("Typical layout setup")
	})

	it("Typical graph setup works", async () => {
		const graph = new Graph(nodes, edges, { layout: { updateCap: Infinity } })
		graph.addLayoutComponent("collide", new Collision())
		graph.addLayoutComponent("manybody", new NBody())
		graph.addLayoutComponent("x", new Attraction(true))
		graph.addLayoutComponent("y", new Attraction(false))
		graph.addLayoutComponent("link", new Link())
		let updateCount = 0
		graph.on("layoutupdate", () => {
			updateCount++
		})
		let resolve
		const promise = new Promise(promiseResolution => (resolve = promiseResolution))
		graph.on("layoutloopend", () => {
			resolve()
		})
		graph.startLayoutLoop()
		await promise
		expect(updateCount).toEqual(300)
		expect(nodes).toMatchSnapshot("Typical graph setup")
	})
})
