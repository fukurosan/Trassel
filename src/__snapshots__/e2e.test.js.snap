// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`End-to-End Tests Typical graph setup works: Typical graph setup 1`] = `
Array [
  Object {
    "id": "n1",
    "index": 0,
    "mass": 2000,
    "radius": 50,
    "vx": -0.0014056788650674168,
    "vy": -0.001027349853782015,
    "x": 126.34726982437574,
    "y": 48.48868792870981,
  },
  Object {
    "id": "n2",
    "index": 1,
    "mass": 2000,
    "radius": 50,
    "vx": -0.0008660706002616717,
    "vy": -0.00025023407573761806,
    "x": -339.068883423034,
    "y": 98.12782225455807,
  },
  Object {
    "id": "n3",
    "index": 2,
    "mass": 2000,
    "radius": 50,
    "vx": -0.002108672191582527,
    "vy": -0.0008286681327984415,
    "x": -64.67406904632774,
    "y": -367.06096116777854,
  },
  Object {
    "id": "n4",
    "index": 3,
    "mass": 2000,
    "radius": 50,
    "vx": 0.0031006504752121578,
    "vy": -0.0029432544464397647,
    "x": 289.0038885379553,
    "y": 466.9373364446614,
  },
  Object {
    "id": "n5",
    "index": 4,
    "mass": 2000,
    "radius": 50,
    "vx": -0.0013747906559529789,
    "vy": -0.00019841500001511266,
    "x": -307.4881314656687,
    "y": -351.0570503748456,
  },
  Object {
    "id": "n6",
    "index": 5,
    "mass": 2000,
    "radius": 50,
    "vx": -0.0012290868613614842,
    "vy": -0.0012607292172967545,
    "x": 151.74595519650606,
    "y": -405.21869156644584,
  },
  Object {
    "id": "n7",
    "index": 6,
    "mass": 2000,
    "radius": 50,
    "vx": -0.004871490431075796,
    "vy": -0.0011855732431741233,
    "x": -215.74443253502764,
    "y": 492.86722036355036,
  },
  Object {
    "id": "n8",
    "index": 7,
    "mass": 2000,
    "radius": 50,
    "vx": 0.0010745053125946632,
    "vy": -0.002354384748288545,
    "x": 9.724152990243976,
    "y": -119.4920128731314,
  },
  Object {
    "id": "n9",
    "index": 8,
    "mass": 2000,
    "radius": 50,
    "vx": 0.0006541199396378107,
    "vy": -0.0014981110164009276,
    "x": 440.8311074156422,
    "y": 32.59643616394459,
  },
  Object {
    "id": "n10",
    "index": 9,
    "mass": 2000,
    "radius": 50,
    "vx": 0.004192713231556837,
    "vy": 0.004443072298939321,
    "x": 42.29660136097145,
    "y": 265.95541791164135,
  },
]
`;

exports[`End-to-End Tests Typical layout loop works: Typical layout setup 1`] = `
Array [
  Object {
    "id": "n1",
    "index": 0,
    "mass": 2000,
    "radius": 50,
    "vx": -0.0014056788650674168,
    "vy": -0.001027349853782015,
    "x": 126.34726982437574,
    "y": 48.48868792870981,
  },
  Object {
    "id": "n2",
    "index": 1,
    "mass": 2000,
    "radius": 50,
    "vx": -0.0008660706002616717,
    "vy": -0.00025023407573761806,
    "x": -339.068883423034,
    "y": 98.12782225455807,
  },
  Object {
    "id": "n3",
    "index": 2,
    "mass": 2000,
    "radius": 50,
    "vx": -0.002108672191582527,
    "vy": -0.0008286681327984415,
    "x": -64.67406904632774,
    "y": -367.06096116777854,
  },
  Object {
    "id": "n4",
    "index": 3,
    "mass": 2000,
    "radius": 50,
    "vx": 0.0031006504752121578,
    "vy": -0.0029432544464397647,
    "x": 289.0038885379553,
    "y": 466.9373364446614,
  },
  Object {
    "id": "n5",
    "index": 4,
    "mass": 2000,
    "radius": 50,
    "vx": -0.0013747906559529789,
    "vy": -0.00019841500001511266,
    "x": -307.4881314656687,
    "y": -351.0570503748456,
  },
  Object {
    "id": "n6",
    "index": 5,
    "mass": 2000,
    "radius": 50,
    "vx": -0.0012290868613614842,
    "vy": -0.0012607292172967545,
    "x": 151.74595519650606,
    "y": -405.21869156644584,
  },
  Object {
    "id": "n7",
    "index": 6,
    "mass": 2000,
    "radius": 50,
    "vx": -0.004871490431075796,
    "vy": -0.0011855732431741233,
    "x": -215.74443253502764,
    "y": 492.86722036355036,
  },
  Object {
    "id": "n8",
    "index": 7,
    "mass": 2000,
    "radius": 50,
    "vx": 0.0010745053125946632,
    "vy": -0.002354384748288545,
    "x": 9.724152990243976,
    "y": -119.4920128731314,
  },
  Object {
    "id": "n9",
    "index": 8,
    "mass": 2000,
    "radius": 50,
    "vx": 0.0006541199396378107,
    "vy": -0.0014981110164009276,
    "x": 440.8311074156422,
    "y": 32.59643616394459,
  },
  Object {
    "id": "n10",
    "index": 9,
    "mass": 2000,
    "radius": 50,
    "vx": 0.004192713231556837,
    "vy": 0.004443072298939321,
    "x": 42.29660136097145,
    "y": 265.95541791164135,
  },
]
`;
